geth/jail/internal/timers/timers.go

```diff
@@ -9,62 +9,82 @@ import (
 	"github.com/status-im/status-go/geth/jail/internal/vm"
 )
 
-var minDelay = map[bool]int64{
-	true:  10,
-	false: 4,
-}
+type intervalFunc func(*loop.Loop) func(call otto.FunctionCall) otto.Value
 
-//Define jail timers
+// Define jail timers
 func Define(vm *vm.VM, l *loop.Loop) error {
 	if v, err := vm.Get("setTimeout"); err != nil {
 		return err
 	} else if !v.IsUndefined() {
 		return nil
 	}
 
-	newTimer := func(interval bool) func(call otto.FunctionCall) otto.Value {
-		return func(call otto.FunctionCall) otto.Value {
-			delay, _ := call.Argument(1).ToInteger()
-			if delay < minDelay[interval] {
-				delay = minDelay[interval]
-			}
-
-			t := &timerTask{
-				duration: time.Duration(delay) * time.Millisecond,
-				call:     call,
-				interval: interval,
-			}
-
-			// If err is non-nil, then the loop is closed and should not
-			// be used anymore.
-			if err := l.Add(t); err != nil {
-				return otto.UndefinedValue()
-			}
-
-			t.timer = time.AfterFunc(t.duration, func() {
-				l.Ready(t) // nolint: errcheck
-			})
-
-			value, newTimerErr := call.Otto.ToValue(t)
-			if newTimerErr != nil {
-				panic(newTimerErr)
-			}
-
-			return value
+	if err := vm.Set("setInterval", newTimer(true, l)); err != nil {
+		return err
+	}
+	if err := vm.Set("setTimeout", newTimer(false, l)); err != nil {
+		return err
+	}
+
+	timers := map[string]intervalFunc{
+		"setImmediate":   immediateTimer,
+		"clearTimeout":   clearTimeout,
+		"clearInterval":  clearTimeout,
+		"clearImmediate": clearTimeout,
+	}
+
+	for k, timer := range timers {
+		if err := vm.Set(k, timer(l)); err != nil {
+			return err
 		}
 	}
 
-	err := vm.Set("setTimeout", newTimer(false))
-	if err != nil {
-		return err
+	return nil
+}
+
+func getDelayWithMin(call otto.FunctionCall, interval bool) int64 {
+	var minDelay = map[bool]int64{
+		true:  10,
+		false: 4,
 	}
 
-	err = vm.Set("setInterval", newTimer(true))
-	if err != nil {
-		return err
+	delay, _ := call.Argument(1).ToInteger()
+	if delay < minDelay[interval] {
+		return minDelay[interval]
 	}
+	return delay
+}
+
+func newTimer(interval bool, l *loop.Loop) func(call otto.FunctionCall) otto.Value {
```

Comment:
A little bit confusing `newTimer` returns some function type.

Reply:
Especially as the signature is defined as type `intervalFunc`. So maybe better

```
func newIntervalTimer(...) intervalFunc { ... }
```

Reply:
Hey @themue that was the idea but i got a compilation error when i tried it. I'll try to do another iteration on how this functions work. As @JekaMas pointed, they look a bit confusing ATM :P

Reply:
Will check the branch and see why it has troubles with the signature.

Reply:
So the newTimer (and other methods) return type `func(call otto.FunctionCall) otto.Value`, which is not of type intervalFunc `func(*loop.Loop) func(call otto.FunctionCall) otto.Value`.

In case we want a clear code we could define a new type like:
```
// New type foo
type foo func(call otto.FunctionCall) otto.Value
// intervalFunc modified to return new type foo
type intervalFunc func(*loop.Loop) foo
// And the newTimer modified to return foo type too
func newTimer(interval bool, l *loop.Loop) foo {
  ...
}
```
However, even this compiles, i get a ton of errors from the tests like:
```
Error: Received unexpected error:		
RangeError: expected 1 argument(s); got 2
```
I have the impression, when compiler calls this function from [here](https://github.com/status-im/status-go/pull/535/files#diff-d5647d5a44b6c6b515fca5f80dd21e7bR22) instead of sending those 2 arguments to newTimer is sending them to the returned [function](https://github.com/status-im/status-go/pull/535/files#diff-d5647d5a44b6c6b515fca5f80dd21e7bR59). That causes that error.

It shouldn't behave like this, but it's what it seems to be doing as far as i can see. 

If you don't like a function returning another funcion (kind of inception movie), we could easily move it to store that function inside a variable as it was in the beginning. But instead of having those variables inside Define function move them outside, so code climate doesn't get angry.

Reply:
An alternative solution would be to get rid of `intervalFunc` type (is it used only once or am I blind?)

The map can be defined as `map[string]func(call otto.FunctionCall) otto.Value`. Finally, we would have:
* `func newTimer(l *loop.Loop, interval bool) func(call otto.FunctionCall) otto.Value`,
* `func newImmediateTimer(l *loop.Loop) func(call otto.FunctionCall) otto.Value`
etc.

It can all be written like this then:
```go
map[string]func(call otto.FunctionCall) otto.Value{
  "setTimeout": newTimer(l, false),
  "setInterval": newTimer(l, true),
  "setImmediate": newImmediateTimer(l),
}

for name, fn := range timers {
  if err := vm.Set(name, fn); err != nil {
    return err
  }
}
```

Or you can go even more "functional" and write `newTimer` like this:
```go
func newTimer(interval bool) func(l *loop) func(call otto.FunctionCall) otto.Value {
  return func (l *loop) func(call otto.FunctionCall) otto.Value {
    // ...
  }
}
```

And then the map will look like this:
```go
map[string]func(call otto.FunctionCall) otto.Value{
  "setTimeout": newTimer(false),
  "setInterval": newTimer(true),
  "setImmediate": newImmediateTimer,
}

for name, fn := range timers {
  if err := vm.Set(name, fn(l)); err != nil {
    return err
  }
}
```

(In such a case, the type `type timerFnConstructor func(l *loop) func(call otto.FunctionCall) otto.Value)` might be actually useful.)

Regarding names, I agree it may be confusing. Something like `newTimerHandler` or `newTimerFunc` may be better.

And replacing `func(call otto.FunctionCall) otto.Value` with a custom type may not work because otto uses reflection to figure out the type and JavaSript arguments. I am not sure how well it's written and if it can figure out recursively the actual type.

Reply:
Yay! all changes are implemented, I hope everybody is happy with the result :)